Improve well typedness rules (#418)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

  to have a result type of `LogicalType`, `SingleNodeType`, or `NodesType`.
* If it occurs directly as a `comparable` in a comparison, the
  function is declared to have a result type of `ValueType` or `SingleNodeType`.
* Each argument of the function conforms to the declared type of the corresponding declared

>   parameter according to one of the following scenarios:

Reviewer: "scenarios" is a bit vague. What's wrong with "rules"?

Author: "Rules" contain a scenario and an outcome.  "If X, then Y" is a rule.  "X" is a scenario, and "Y" is the outcome that applies when that scenario exists.

> The declared type of the parameter is `ValueType` and the argument is one of...

is a scenario.  The outcome is indicated above:

> Each argument of the function conforms to the declared type of the corresponding declared
  parameter

Those _together_ are a rule.

Author: We _could_ reword this slightly to be a rule if you prefer:

>  * If the declared type of the parameter is `ValueType`, then the argument must be one of:
>    * a function with declared type of `ValueType` or `SingleNodeType`
>    * a `singular-path`
>    * a `literal`

... etc.

Reviewer: I get it now, thanks. Actually I suggest:
```suggestion
  parameter according to one of the following:
```

   * The declared type of the parameter is `ValueType` and the argument is one of:
     * a function with declared type of `ValueType` or `SingleNodeType`

>      * a `singular-path`

Reviewer: The PR needs to explain how to get from `SingleNodeType` and `singular-path` to `ValueType`. C.f. the following text from the "Comparisons" section:
> When any path on either side of a comparison results in a nodelist consisting of a single node, each such path is replaced by the value of its node

Author: The text on this branch actually reads:

> When any path **or function expression** on either side of a comparison results in a nodelist consisting of a single node, each such path **or function expression** is replaced by the value of its node and then...

so it's already there.

Reviewer: Certainly that text is the kind of thing that is needed. But that text limits its application to either side of a comparison and this excludes arguments passed as parameters of type `ValueType`.

Author: Addressed in [ba8c16b](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/418/commits/ba8c16bf51f0aa495c605ca16e46006a1dbb3340)

Reviewer: Yes, that does the job nicely.

     * a `literal`
   * The declared type of the parameter is `LogicalType` and the argument is one of:
     * a function with declared type of `LogicalType`, `SingleNodeType`, or `NodesType`

>      * a `filter-path` (which includes `singular-path`)

Reviewer: The PR needs to explain how to get from `SingleNodeType`, `NodesType`, and `filter-path` to `LogicalType`. C.f. the following text from "Existence Tests":
> A path by itself in a Logical context is an existence test which yields true if the path selects at least one node and yields false if the path does not select any nodes.

Author: The text on this branch actually reads:

> A path by itself in a Logical context is an existence test which yields true if the path selects at least one node and yields false if the path does not select any nodes.
> A function expression with result type `SingleNodeType` or `NodesType` in a Logical context is an existence test which yields true if the function returns a non-empty nodelist and yields false if the function returns an empty node-list.

so it's already there.

Author: > supplying an argument to a parameter with type `LogicalType` has any connection with a "logical context"

I don't understand.  How can anywhere `LogicalType` is used _not_ be a "logical context?"

Reviewer: Because we haven't actually defined "logical context", so it's somewhat open to interpretation. My interpretation in the context of the existence test section is that "logical context" refers to "logical-expr":
```
logical-expr        = logical-or-expr
logical-or-expr     = logical-and-expr *(S "||" S logical-and-expr)
                        ; disjunction
                        ; binds less tightly than conjunction
logical-and-expr    = basic-expr *(S "&&" S basic-expr)
                        ; conjunction
                        ; binds more tightly than disjunction

basic-expr          = paren-expr /
                      comparison-expr /
                      test-expr

```

Author: Addressed in [ba8c16b](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/418/commits/ba8c16bf51f0aa495c605ca16e46006a1dbb3340)

Reviewer: Yes, that does the job nicely.

  to have a result type of `LogicalType`, `SingleNodeType`, or `NodesType`.
* If it occurs directly as a `comparable` in a comparison, the
  function is declared to have a result type of `ValueType` or `SingleNodeType`.
* Each argument of the function conforms to the declared type of the corresponding declared
  parameter according to one of the following scenarios:
   * The declared type of the parameter is `ValueType` and the argument is one of:
     * a function with declared type of `ValueType` or `SingleNodeType`
     * a `singular-path`
     * a `literal`
   * The declared type of the parameter is `LogicalType` and the argument is one of:
     * a function with declared type of `LogicalType`, `SingleNodeType`, or `NodesType`
     * a `filter-path` (which includes `singular-path`)
   * The declared type of the parameter is `SingleNodeType` and the argument is one of:
     * a function with declared type of `SingleNodeType`
     * a `singular-path`
   * The declared type of the parameter is `NodesType` and the argument is one of:

>      * a function with declared type of `SingleNodeType` or `NodesType`

Reviewer: The other scenarios treat the parameter type first, so:
```suggestion
     * a function with declared type of `NodesType` or `SingleNodeType`
```

Author: I was actually listing them in the order they're in the table.

Reviewer: I'm afraid that subtlety was lost on me.

Author: I'm not sure it really matters.

Reviewer: Let's not waste more time on editorial niceties then.

     * a `filter-path`

### Processing Nodelists as Arguments

> When any `singular-path` or function expression with declared type of `SingleNodeType` appears as an argument for a parameter of type `ValueType`, each such path or function expression is replaced by the value of its node or `Nothing` if it is empty.

Reviewer: There's an editorial wrinkle in that `Nothing` isn't really empty. How about the following wordier, but hopefully clearer possibility? If you don't like this, I'd be happy to apply it as an editorial tweak if this PR is merged.
```suggestion
When any `singular-path` appears as an argument for a parameter of type `ValueType`, the path is replaced by the value of the node in the resulting nodelist or `Nothing` if the resulting nodelist is empty.

When any function expression with declared type of `SingleNodeType` appears as an argument for a parameter of type `ValueType`, the function expression is replaced by the value of the node in the resulting nodelist or `Nothing` if the function expression returns `Nothing`.
```

Author: Maybe this tongue-in-cheek kebaberizing will help:

> When any `singular-path` or function-expression-with-declared-type-of-`SingleNodeType` appears as an argument for a parameter of type `ValueType`, each such path or function expression is replaced by the value of its node or `Nothing` if it is empty.

I agree that it's a poor result of English grammar.  Maybe if I reverse them:

> When any function expression with declared type of `SingleNodeType` or `singular-path` appears as an argument for a parameter of type `ValueType`, each such path or function expression is replaced by the value of its node or `Nothing` if it is empty.

It still has the grammatical issue, but since a function can't result in a `singular-path`, maybe it makes more sense.

Reviewer: Let's not waste more time on editorial niceties then.
